### PR TITLE
[FSDP2] Fix tied weights with uneven sharding across separate FSDP groups

### LINKED ISSUE #176076: A portion of shared parameter is not being update when sharding is uneven over FSDP2
### Summary

There is a subtle bug in FSDP2 that occurs under the following conditions:

* FSDP2 is used
* Parameters are shared across modules (e.g., tied embeddings)
* Parameter sizes are uneven across ranks (e.g., world size does not divide vocab size)

When all three conditions are met, the parameters on the **last rank are never updated** during training.

This can be reproduced when training a model with tied embeddings (e.g., Qwen-style tied input/output embeddings) and using a number of FSDP ranks that does not evenly divide the vocabulary size. Only the parameters on the last rank are affected; all other ranks update normally.

---

## Minimal Reproduction

The following script reproduces the issue:

```python
import torch.distributed.checkpoint
import torch.distributed.fsdp
import torch
import os

torch.distributed.init_process_group(backend="nccl")
torch.set_printoptions(precision=10, sci_mode=False, profile="full")

local_rank = int(os.environ["LOCAL_RANK"])
print(f"Running on local rank: {local_rank}")
torch.cuda.set_device(local_rank)
device = torch.device(f"cuda:{local_rank}")

mesh = torch.distributed.init_device_mesh(
    "cuda",
    (torch.distributed.get_world_size(),),
    mesh_dim_names=("fsdp",)
)

torch.manual_seed(42 + torch.distributed.get_rank())
torch.cuda.manual_seed(42 + torch.distributed.get_rank())

hidden_size = 16
sequence_length = 8
batch_size = 2
vocab_size = 17  # Not divisible by world size

ids = torch.arange(batch_size * sequence_length, device=device).view(batch_size, sequence_length) % vocab_size
labels = torch.arange(batch_size * sequence_length, device=device).view(batch_size, sequence_length) % vocab_size

class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.tok_embeddings = torch.nn.Embedding(vocab_size, hidden_size)
        self.output = torch.nn.Linear(hidden_size, vocab_size, bias=False)

        # Weight tying
        self.output.weight = self.tok_embeddings.weight

    def forward(self, ids):
        x = self.tok_embeddings(ids)
        return self.output(x)

    def init_weights(self):
        torch.nn.init.normal_(self.tok_embeddings.weight, mean=0.0, std=0.02)
        self.output.weight = self.tok_embeddings.weight

model = Model()

model.tok_embeddings = torch.distributed.fsdp.fully_shard(
    model.tok_embeddings, mesh=mesh, reshard_after_forward=True
)
model.output = torch.distributed.fsdp.fully_shard(
    model.output, mesh=mesh, reshard_after_forward=True
)
model = torch.distributed.fsdp.fully_shard(
    model, mesh=mesh, reshard_after_forward=True
)

model.to_empty(device="cuda")
with torch.no_grad():
    model.init_weights()

model.tok_embeddings.unshard()
if torch.distributed.get_rank() == 3:
    print(model.tok_embeddings.weight[:2])
    print(model.tok_embeddings.weight[-2:])
model.tok_embeddings.reshard()

optimizer = torch.optim.AdamW(model.parameters(), lr=1e-2)

for _ in range(20):
    logits = model(ids)
    loss = torch.nn.functional.cross_entropy(
        logits.view(-1, vocab_size).float(),
        labels.view(-1)
    )
    loss.backward()
    optimizer.step()
    optimizer.zero_grad()

model.tok_embeddings.unshard()
if torch.distributed.get_rank() == 3:
    print(model.tok_embeddings.weight[:2]) # params are updated
    print(model.tok_embeddings.weight[-2:]) # params not updated
model.tok_embeddings.reshard()
```

After training, the last rank’s shard remains unchanged.

---

## Suspected Root Cause

The issue appears to originate in the padding logic used to equalize shard sizes across ranks:

[https://github.com/pytorch/pytorch/blob/7e9cf0e7c069e7d6150df9e79fa47294bbbc88ed/torch/distributed/fsdp/_fully_shard/_fsdp_param.py#L855-L920](https://github.com/pytorch/pytorch/blob/7e9cf0e7c069e7d6150df9e79fa47294bbbc88ed/torch/distributed/fsdp/_fully_shard/_fsdp_param.py#L855-L920)

As I understand it, this code re-pads local tensors to ensure all ranks have equal shard sizes so that FSDP2 all-gather operations can proceed correctly.

The problem arises when this logic runs after weight tying has been established.

### What Happens

When calling `model(ids)` in the reproduction script:

* Two `FSDPParam` instances are being managed:

  * One for the embedding layer
  * One for the output layer

1. `reset_sharded_param` is triggered for the embedding layer.

   * On the last rank, `self._sharded_param_data` is reassigned to a padded tensor.

2. `reset_sharded_param` is then triggered for the output layer.

   * On the last rank, `self._sharded_param_data` is again reassigned to a padded tensor.

However, because the embedding and output layers share the same underlying parameter (due to weight tying), reassigning `self._sharded_param_data` for one `FSDPParam` effectively changes the storage backing the shared parameter.

At this point:

* The internal storage for the embedding layer has changed.
* But `self.sharded_param._local_tensor` for the embedding layer is not updated accordingly.

This leads to the embedding’s local tensor pointing to stale storage. As a result, optimizer updates do not affect the actual parameter data for the last rank.

### Versions

No entry for terminal type "alacritty";
using dumb terminal settings.
Collecting environment information...
PyTorch version: 2.10.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux release 8.8 (Ootpa) (x86_64)
GCC version: (GCC) 8.5.0 20210514 (Red Hat 8.5.0-18)
Clang version: Could not collect
CMake version: version 3.20.2
Libc version: glibc-2.28

Python version: 3.13.5 (main, Jun 12 2025, 12:40:22) [Clang 20.1.4 ] (64-bit runtime)
Python platform: Linux-4.18.0-477.27.1.el8_8.x86_64-x86_64-with-glibc2.28
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              128
On-line CPU(s) list: 0-127
Thread(s) per core:  2
Core(s) per socket:  32
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               106
Model name:          Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
Stepping:            6
CPU MHz:             2601.000
CPU max MHz:         2601.0000
CPU min MHz:         800.0000
BogoMIPS:            5200.00
Virtualization:      VT-x
L1d cache:           48K
L1i cache:           32K
L2 cache:            1280K
L3 cache:            49152K
NUMA node0 CPU(s):   0-31,64-95
NUMA node1 CPU(s):   32-63,96-127
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @xmfan

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #176225

Fix _validate_no_duplicate_params false positive from id() reuse                                                                                       
                                                                                                                                                         
  _validate_no_duplicate_params detects tied weights placed in separate                                                                                  
  FSDP groups by comparing id(param) values captured at FSDPParam
  creation time. However, Python reuses id() values for
  garbage-collected objects. In layer-by-layer init patterns (materialize
  block, fully_shard, GC, next block), freed parameters' id() values
  get reused by new parameters in later blocks, triggering a spurious
  "already managed by another FSDP group" error.

  The fix replaces id(param) with a monotonically increasing UID stamped
  on the parameter object itself. For tied weights, both modules share the
  same nn.Parameter object, so they see the same UID and the duplicate
  check still fires correctly. For different parameters that happen to
  recycle the same id() after GC, each gets its own fresh UID.

  Add a test that materializes and shards blocks one at a time with
  gc.collect() between each, verifying that forward and backward
  complete without a false-positive duplicate parameter error.

Fixes #176076

Authored with Claude.

### CODE REVIEW

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 1]
```diff
@@ -1029,6 +1029,73 @@ def _test_train_shared_params(
...
+                return self.output(self.tok_embeddings(ids))
+
+        torch.manual_seed(42)
+        model = TiedModel()
+        fully_shard(model.tok_embeddings, reshard_after_forward=reshard_after_forward)
+        fully_shard(model.output, reshard_after_forward=reshard_after_forward)
+        fully_shard(model, reshard_after_forward=reshard_after_forward)
+
+        # Re-establish weight tying (broken by separate fully_shard calls)
+        model.output.weight = model.tok_embeddings.weight
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py]: those 2 lines are still working around of invalid fully_shard wrapping

2 alternatives seems to be reasonable
* allow fully_shard(model.tok_embeddings) and fully_shard(model.output), no need to call `model.output.weight = model.tok_embeddings.weight`, unshard/reshard
* throw error when applying fully_shard(model.tok_embeddings) and fully_shard(model.output). We only allow fully_shard([model.tok_embeddings, model.output])

Comment (francesco-bertolotti) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py]: I didn’t realize that `fully_shard([model.tok_embeddings, model.output])` was even supported.

If this is the intended way to use FSDP2 when weights are shared, should I assume that doing

```python
fully_shard(model.tok_embeddings)
fully_shard(model.output)
```

would lead to issues?

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py [hunk 2]
```diff
@@ -802,12 +804,26 @@ def reset_sharded_param(self):
                 raise AssertionError(
                     f"Shard({shard_dim}) requires even sharding: {local_tensor.size()=}"
                 )
-            padded_local_tensor = local_tensor.new_zeros(padded_sharded_size)
-            padded_local_tensor.narrow(dim=shard_dim, start=0, length=length).copy_(
-                local_tensor
+            # Check if local_tensor is already a narrow view of a padded buffer
```
Comment (weifengpy) [inline torch/distributed/fsdp/_fully_shard/_fsdp_param.py]: are line 807 - 826 obsolete now? we are raising hard error to let user wrap tiled weights together

those lines are only relevant later if we decide to suport wrapping tiled weights separately

Comment (wconstab) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param.py]: removing!

**Review by weifengpy — APPROVED** [review]
approve. probably need to update the unit test to avoid ci error
